### PR TITLE
Feat/return simulation result as object

### PR DESCRIPTION
**Pull request by miguel-bm**
#### Description

This feature will enable us to carry out multiple experiments and aggregate the results with ease, without having to write and read multiple result files. All of the results currently written into files are now also returned as an object by the `run_simulation` method.

### CODE REVIEW

File: simulator/utilities/logger.py [hunk 1]
```diff
@@ -39,7 +40,7 @@ def set_write_events(self, write_events: bool):
         if not write_events:
             self.event_logger.handlers = []
 
-    def set_filename(self, filename: str):
+    def set_filename(self, filename: Union[Path, str]):
         self.action_logger.handlers = []
```
Comment (miguel-bm) [inline simulator/utilities/logger.py:44]: Boyscouting.

File: simulator/state/state.py [hunk 2]
```diff
@@ -35,24 +35,26 @@ def __init__(
...
         self._tick += 1
         self._time += time_step
 
         # The environment advances, updating time and prices
-        self._environment.take_step(time_step)
+        environment_events_info = self._environment.take_step(time_step)
 
         # The protocol advances, some automatic actions might take place
-        self._protocol.take_step(time_step)
+        protocol_events_info = self._protocol.take_step(time_step)
```
Comment (miguel-bm) [inline simulator/state/state.py:47]: Now these `take_step` methods return optional info about what happened during the step so that it can be reported to the results.

File: simulator/state/state.py [hunk 3]
```diff
@@ -35,24 +35,26 @@ def __init__(
...
         self._agents = self._strategy.update_agents(
             self._agents, self._protocol, self._environment, time_step
         )
 
         # Agents act depending on the protocol state and environment, modifying them.
+        actions_info: List[dict] = []
         for agent in self._agents_by_priority():
-            agent.act()
+            actions_info.extend(agent.act())  # TODO: agent.act()
+        return actions_info, environment_events_info + protocol_events_info
```
Comment (miguel-bm) [inline simulator/state/state.py:57]: Now the `act` method should return the list of actions that the agent executed within this step, so that it can be reported to the results.

File: simulator/result/result.py [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+    write_step_metrics: bool = True
+    write_end_metrics: bool = True
+    write_custom_event_metrics: bool = True
+    write_log: bool = True
+    meta_log_level: str = "INFO"
+    step_log_interval: int = 1
+
+
+class SimulationResultWriter:
+    def __init__(self, run_id: str, config: SimulationResultConfig):
```
Comment (miguel-bm) [inline simulator/result/result.py:34]: Main new functionality introduced.

File: simulator/result/result.py [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+    def add_end_metrics(self, end_metrics: dict):
+        self.result.end_metrics = end_metrics
+        if self.config.write_end_metrics:  # TODO: JSONify
+            self.storage.write_or_append_csv("end_metrics", [end_metrics])
+
+    def add_action_to_log(self, action_info: dict):
+        self.result.action_log.append(action_info)
+        message = action_info["message"]  # TODO: Redo this shit
+        del action_info["message"]
+        logger.log_action(message, action_info)
```
Comment (miguel-bm) [inline simulator/result/result.py:71]: Very clunky and don't like this. Probably better to redo the `action_info` as a `dataclass` with the appropriate fields, or simply redo how the actions are logged. WDYT @crissto ?

File: simulator/metrics/base_metrics.py [hunk 6]
```diff
@@ -11,6 +12,11 @@ class BaseSimulationMetrics:
     def by_step(cls, state: SimulationState) -> Metric:
         pass
 
+    @classmethod
+    @abstractmethod
+    def custom_events(cls, state: SimulationState) -> Dict[str, List[dict]]:
+        pass
+
```
Comment (miguel-bm) [inline simulator/metrics/base_metrics.py:19]: New type of metric we should consider. In the **cora** case this would be things like a cycle completion, when we could calculate specific stats about the cycle.

File: simulator/state/state.py (hunk 2 above)
Comment (miguel-bm) [reply simulator/state/state.py:47]: Should probably make the return types `List[EventInfo]` or some kind of DTO like that.

File: simulator/engine/engine.py [hunk 7]
```diff
@@ -1,26 +1,20 @@
...
-from itertools import count
-
 
-@dataclass
-class SimulationConfig:
-    results_folder: str = "simlogs"
-    write_tick_metrics: bool = True
-    write_end_metrics: bool = True
-    write_action_log: bool = True
-    meta_log_level: str = "INFO"
```
Comment (miguel-bm) [inline simulator/engine/engine.py:23]: Moved to `result.py`

File: simulator/engine/engine.py [hunk 8]
```diff
@@ -90,6 +68,24 @@ def run_simulation(
...
+    def run_simulations(
+        self,
+        num_runs: int,
+        start: datetime,
+        end: datetime,
+        step: timedelta,
+        random_seed: int = 513,
+    ) -> List[SimulationResult]:
+
+        random_seeds = [random_seed * (i + 1) for i in range(num_runs)]
```
Comment (miguel-bm) [inline simulator/engine/engine.py:82]: I know this is not very good... But works for now.

File: protocols/cora/v1/tests/test_metrics.py [hunk 9]
```diff
@@ -133,9 +133,9 @@ def filter_metrics(metric: dict, key: Union[str, Pattern]):
...
 
 def test_sum_of_loans():
-    engine = run_simulation()
+    result = run_simulation()
 
-    for metric in engine._metric_history["log"]:
+    for metric in result.step_metrics:
         assert (
             metric["expired_loans_count"]
             == metric["paid_loans_count"] + metric["defaulted_loans_count"]
```
Comment (miguel-bm) [inline protocols/cora/v1/tests/test_metrics.py:141]: These kind of tests now come more naturally after the change I think.

File: protocols/cora/v1/protocol.py [hunk 10]
```diff
@@ -19,22 +18,11 @@ def __init__(self, environment: BaseCoraEnvironment) -> None:
...
-        logger.log_event(
-            f"{current_time}: Taking step of {time_step.seconds} seconds, "
-            f"current price is {current_price:.4f}",
-            {
-                "current_time": current_time.isoformat(),
-                "event": "protocol_step",
-                "time_step": time_step.seconds,
-                "current_price": current_price,
-            },
-        )
```
Comment (miguel-bm) [inline protocols/cora/v1/protocol.py:35]: Redid in the new implementation and moved to the environment, since it is more appropriate there. This log said nothing about the protocol itself.

File: protocols/cora/v1/metrics.py [hunk 11]
```diff
@@ -116,6 +112,10 @@ def by_step(cls, state: SimulationState) -> Metric:
             **binned_metrics,
         }
 
+    @classmethod
+    def custom_events(cls, state: SimulationState) -> Dict[str, List[dict]]:
+        return {}
+
```
Comment (miguel-bm) [inline protocols/cora/v1/metrics.py:118]: No implementation of these new metrics for now, will be in a separate PR.

**Review by crissto — APPROVED** [review]
Niiiice
